File: src/native_python_processors/ReceiveFromNodeJSApp/ReceiveFromNodeJSApp.py
import json
import base64
import urllib.request
import urllib.parse
import urllib.error
from nifiapi.flowfiletransform import FlowFileTransform, FlowFileTransformResult
from nifiapi.properties import PropertyDescriptor, StandardValidators, ExpressionLanguageScope

class ReceiveFromNodeJSApp(FlowFileTransform):
    class Java:
        implements = ['org.apache.nifi.python.processor.FlowFileTransform']

    class ProcessorDetails:
        version = '1.0.0-SNAPSHOT'
        description = '''Receives HTTP requests from Node.js applications via the NodeJSAppAPIGateway controller service.
        This processor polls the gateway's internal API for requests matching the configured endpoint pattern.'''
        tags = ['nodejs', 'http', 'gateway', 'api', 'rest', 'python']

    def __init__(self, **kwargs):
        self.gateway_url = None
        self.endpoint_pattern = None
        self.poll_timeout = 30

    GATEWAY_URL = PropertyDescriptor(
        name="Gateway URL",
        description="Base URL of the NodeJSAppAPIGateway service (e.g., http://localhost:5050)",
        required=True,
        default_value="http://localhost:5050",
        validators=[StandardValidators.NON_EMPTY_VALIDATOR],
        expression_language_scope=ExpressionLanguageScope.ENVIRONMENT
    )

    ENDPOINT_PATTERN = PropertyDescriptor(
        name="Endpoint Pattern",
        description="The endpoint pattern to poll for (e.g., /api/events, /users/:userId). Must match a pattern registered with the gateway.",
        required=True,
        validators=[StandardValidators.NON_EMPTY_VALIDATOR],
        expression_language_scope=ExpressionLanguageScope.ENVIRONMENT
    )

    POLL_TIMEOUT = PropertyDescriptor(
        name="Poll Timeout",
        description="Timeout in seconds for long-polling requests to the gateway",
        required=True,
        default_value="30",
        validators=[StandardValidators.POSITIVE_INTEGER_VALIDATOR]
    )

    def getPropertyDescriptors(self):
        return [self.GATEWAY_URL, self.ENDPOINT_PATTERN, self.POLL_TIMEOUT]

    def onScheduled(self, context):
        """Called when the processor is scheduled to run."""
        self.gateway_url = context.getProperty(self.GATEWAY_URL).evaluateAttributeExpressions().getValue()
        self.endpoint_pattern = context.getProperty(self.ENDPOINT_PATTERN).evaluateAttributeExpressions().getValue()
        self.poll_timeout = int(context.getProperty(self.POLL_TIMEOUT).getValue())

        # Remove trailing slash from gateway URL if present
        if self.gateway_url.endswith('/'):
            self.gateway_url = self.gateway_url[:-1]

    def transform(self, context, flowFile):
        """
        Polls the gateway for the next request matching the endpoint pattern.
        Creates a new FlowFile with the request data.
        """
        try:
            # Build polling URL
            # URL encode the endpoint pattern
            encoded_pattern = urllib.parse.quote(self.endpoint_pattern, safe='')
            poll_url = f"{self.gateway_url}/_internal/poll/{encoded_pattern}"

            self.logger.debug(f"Polling gateway at: {poll_url}")

            # Create request with timeout
            req = urllib.request.Request(poll_url, method='GET')
            req.add_header('Content-Type', 'application/json')

            # Poll with timeout
            try:
                with urllib.request.urlopen(req, timeout=self.poll_timeout + 5) as response:
                    if response.status == 204:
                        # No content - timeout, no request available
                        self.logger.debug("Poll timeout - no requests available")
                        return FlowFileTransformResult(relationship="success")

                    if response.status != 200:
                        self.logger.error(f"Gateway returned status {response.status}")
                        return FlowFileTransformResult(relationship="failure")

                    # Read and parse response
                    response_data = response.read().decode('utf-8')
                    request_json = json.loads(response_data)

            except urllib.error.HTTPError as e:
                if e.code == 404:
                    self.logger.error(f"Endpoint pattern '{self.endpoint_pattern}' not registered with gateway")
                    return FlowFileTransformResult(relationship="failure")
                else:
                    self.logger.error(f"HTTP error polling gateway: {e.code} - {e.reason}")
                    return FlowFileTransformResult(relationship="failure")

            except urllib.error.URLError as e:
                self.logger.error(f"URL error polling gateway: {str(e.reason)}")
                return FlowFileTransformResult(relationship="failure")

            except TimeoutError:
                self.logger.debug("Poll timeout - no requests available")
                return FlowFileTransformResult(relationship="success")

            # Extract request data
            method = request_json.get('method', '')
            path = request_json.get('path', '')
            headers = request_json.get('headers', {})
            query_params = request_json.get('queryParameters', {})
            path_params = request_json.get('pathParameters', {})
            content_type = request_json.get('contentType', '')
            body_base64 = request_json.get('body', '')
            body_text = request_json.get('bodyText', '')
            client_address = request_json.get('clientAddress', '')
            timestamp = request_json.get('timestamp', '')

            self.logger.info(f"Received request: {method} {path}")

            # Decode body from base64
            if body_base64:
                try:
                    body_bytes = base64.b64decode(body_base64)
                except Exception as e:
                    self.logger.error(f"Failed to decode body: {str(e)}")
                    body_bytes = body_text.encode('utf-8')
            else:
                body_bytes = b''

            # Build attributes
            attributes = {
                'http.method': method,
                'http.path': path,
                'http.client.address': client_address,
                'http.timestamp': timestamp
            }

            # Add query parameters
            for key, value in query_params.items():
                attributes[f'http.query.{key}'] = str(value)

            # Add path parameters
            for key, value in path_params.items():
                attributes[f'http.path.{key}'] = str(value)

            # Add headers
            for key, value in headers.items():
                attributes[f'http.header.{key.lower()}'] = str(value)

            # Create FlowFile with body content and attributes
            return FlowFileTransformResult(
                relationship="success",
                contents=body_bytes,
                attributes=attributes
            )

        except Exception as e:
            self.logger.error(f"Error processing request: {str(e)}")
            import traceback
            self.logger.error(traceback.format_exc())
            return FlowFileTransformResult(relationship="failure")

File: src/java_extensions/nodejs-app-gateway/nodejs-app-gateway-processors/src/main/java/org/nocodenation/nifi/nodejsapp/gateway/ReceiveFromNodeJSApp.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.nocodenation.nifi.nodejsapp.gateway;

import org.apache.nifi.annotation.behavior.InputRequirement;
import org.apache.nifi.annotation.behavior.WritesAttribute;
import org.apache.nifi.annotation.behavior.WritesAttributes;
import org.apache.nifi.annotation.documentation.CapabilityDescription;
import org.apache.nifi.annotation.documentation.Tags;
import org.apache.nifi.annotation.lifecycle.OnScheduled;
import org.apache.nifi.annotation.lifecycle.OnStopped;
import org.apache.nifi.components.PropertyDescriptor;
import org.apache.nifi.expression.ExpressionLanguageScope;
import org.apache.nifi.flowfile.FlowFile;
import org.apache.nifi.processor.AbstractProcessor;
import org.apache.nifi.processor.ProcessContext;
import org.apache.nifi.processor.ProcessSession;
import org.apache.nifi.processor.Relationship;
import org.apache.nifi.processor.exception.ProcessException;
import org.apache.nifi.processor.util.StandardValidators;

import java.util.*;
import java.util.concurrent.atomic.AtomicLong;

/**
 * Processor that receives HTTP requests from Node.js applications via the NodeJSAppAPIGateway.
 *
 * <p>This processor registers one or more endpoint patterns with the gateway and handles incoming
 * requests directly (zero-latency Java handler). Each request becomes a FlowFile with
 * HTTP details as attributes and body as content.</p>
 *
 * <p><b>Endpoint Pattern Examples:</b></p>
 * <ul>
 *   <li><code>/api/events</code> - Single static path</li>
 *   <li><code>/users/:userId</code> - Single path with parameter</li>
 *   <li><code>/api/users/:userId, /api/products/:productId</code> - Multiple comma-separated patterns</li>
 * </ul>
 *
 * <p><b>FlowFile Attributes:</b></p>
 * <ul>
 *   <li><code>http.method</code> - HTTP method (GET, POST, etc.)</li>
 *   <li><code>http.path</code> - Actual request path (e.g., /api/users/123)</li>
 *   <li><code>http.query.*</code> - Query parameters</li>
 *   <li><code>http.path.*</code> - Path parameters extracted from pattern (e.g., http.path.userId=123)</li>
 *   <li><code>http.header.*</code> - HTTP headers</li>
 *   <li><code>http.client.address</code> - Client IP address</li>
 *   <li><code>http.timestamp</code> - Request timestamp</li>
 * </ul>
 *
 * @since 1.0.0
 */
@Tags({"nodejs", "http", "gateway", "api", "rest"})
@CapabilityDescription("Receives HTTP requests from Node.js applications via the NodeJSAppAPIGateway controller service. " +
        "Registers an endpoint pattern and creates FlowFiles for incoming requests with HTTP details as attributes.")
@InputRequirement(InputRequirement.Requirement.INPUT_FORBIDDEN)
@WritesAttributes({
        @WritesAttribute(attribute = "http.method", description = "HTTP method (GET, POST, PUT, DELETE, etc.)"),
        @WritesAttribute(attribute = "http.path", description = "Request path"),
        @WritesAttribute(attribute = "http.query.*", description = "Query parameters (e.g., http.query.id)"),
        @WritesAttribute(attribute = "http.path.*", description = "Path parameters (e.g., http.path.userId)"),
        @WritesAttribute(attribute = "http.header.*", description = "HTTP headers (e.g., http.header.content-type)"),
        @WritesAttribute(attribute = "http.client.address", description = "Client IP address"),
        @WritesAttribute(attribute = "http.timestamp", description = "Request timestamp (ISO-8601)")
})
public class ReceiveFromNodeJSApp extends AbstractProcessor {

    public static final PropertyDescriptor GATEWAY_SERVICE = new PropertyDescriptor.Builder()
            .name("Gateway Service")
            .description("The NodeJSAppAPIGateway controller service to use")
            .required(true)
            .identifiesControllerService(NodeJSAppAPIGateway.class)
            .build();

    public static final PropertyDescriptor ENDPOINT_PATTERN = new PropertyDescriptor.Builder()
            .name("Endpoint Pattern")
            .description("One or more endpoint patterns to register, comma-separated (e.g., /api/events or /api/users/:userId, /api/products/:productId)")
            .required(true)
            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
            .expressionLanguageSupported(ExpressionLanguageScope.ENVIRONMENT)
            .build();

    public static final PropertyDescriptor RESPONSE_STATUS_CODE = new PropertyDescriptor.Builder()
            .name("Response Status Code")
            .description("HTTP status code to return (default: 202 Accepted)")
            .required(true)
            .defaultValue("202")
            .addValidator(StandardValidators.POSITIVE_INTEGER_VALIDATOR)
            .build();

    public static final PropertyDescriptor RESPONSE_BODY = new PropertyDescriptor.Builder()
            .name("Response Body")
            .description("Optional response body to send back to the client")
            .required(false)
            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
            .expressionLanguageSupported(ExpressionLanguageScope.FLOWFILE_ATTRIBUTES)
            .build();

    public static final Relationship REL_SUCCESS = new Relationship.Builder()
            .name("success")
            .description("Successfully received requests")
            .build();

    public static final Relationship REL_FAILURE = new Relationship.Builder()
            .name("failure")
            .description("Failed to process requests")
            .build();

    private static final List<PropertyDescriptor> PROPERTY_DESCRIPTORS;
    private static final Set<Relationship> RELATIONSHIPS;

    static {
        List<PropertyDescriptor> props = new ArrayList<>();
        props.add(GATEWAY_SERVICE);
        props.add(ENDPOINT_PATTERN);
        props.add(RESPONSE_STATUS_CODE);
        props.add(RESPONSE_BODY);
        PROPERTY_DESCRIPTORS = Collections.unmodifiableList(props);

        Set<Relationship> rels = new HashSet<>();
        rels.add(REL_SUCCESS);
        rels.add(REL_FAILURE);
        RELATIONSHIPS = Collections.unmodifiableSet(rels);
    }

    private NodeJSAppAPIGateway gateway;
    private List<String> endpointPatterns;
    private int responseStatusCode;
    private String responseBody;
    private final AtomicLong requestCounter = new AtomicLong(0);

    @Override
    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
        return PROPERTY_DESCRIPTORS;
    }

    @Override
    public Set<Relationship> getRelationships() {
        return RELATIONSHIPS;
    }

    @OnScheduled
    public void onScheduled(ProcessContext context) throws EndpointAlreadyRegisteredException {
        this.gateway = context.getProperty(GATEWAY_SERVICE).asControllerService(NodeJSAppAPIGateway.class);
        String patternsValue = context.getProperty(ENDPOINT_PATTERN).evaluateAttributeExpressions().getValue();
        this.responseStatusCode = context.getProperty(RESPONSE_STATUS_CODE).asInteger();
        this.responseBody = context.getProperty(RESPONSE_BODY).getValue();

        // Parse comma-separated patterns
        this.endpointPatterns = new ArrayList<>();
        for (String pattern : patternsValue.split(",")) {
            String trimmed = pattern.trim();
            if (!trimmed.isEmpty()) {
                endpointPatterns.add(trimmed);
            }
        }

        // Register all endpoint patterns with null handler and custom response configuration
        for (String pattern : endpointPatterns) {
            gateway.registerEndpoint(pattern, null, responseStatusCode, responseBody);
            getLogger().info("Registered endpoint pattern '{}' with gateway on port {} (status: {}, body: {})",
                    pattern, gateway.getGatewayPort(), responseStatusCode,
                    responseBody != null ? "configured" : "default");
        }
    }

    @OnStopped
    public void onStopped() {
        if (gateway != null && endpointPatterns != null) {
            for (String pattern : endpointPatterns) {
                gateway.unregisterEndpoint(pattern);
                getLogger().info("Unregistered endpoint pattern '{}'", pattern);
            }
        }
    }

    @Override
    public void onTrigger(ProcessContext context, ProcessSession session) throws ProcessException {
        // Poll all registered endpoint queues round-robin
        GatewayRequest request = null;
        String matchedPattern = null;

        for (String pattern : endpointPatterns) {
            java.util.Queue<GatewayRequest> queue = gateway.getEndpointQueue(pattern);

            if (queue == null) {
                getLogger().error("No queue found for endpoint pattern '{}'", pattern);
                continue;
            }

            // Cast to BlockingQueue for poll with timeout support
            if (!(queue instanceof java.util.concurrent.BlockingQueue)) {
                getLogger().error("Queue for pattern '{}' does not support blocking operations", pattern);
                continue;
            }

            java.util.concurrent.BlockingQueue<GatewayRequest> blockingQueue =
                (java.util.concurrent.BlockingQueue<GatewayRequest>) queue;

            // Poll with short timeout
            try {
                request = blockingQueue.poll(100, java.util.concurrent.TimeUnit.MILLISECONDS);
                if (request != null) {
                    matchedPattern = pattern;
                    break; // Found a request, process it
                }
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                getLogger().warn("Interrupted while polling for requests");
                return;
            }
        }

        if (request == null) {
            // No request available from any queue, yield
            context.yield();
            return;
        }

        try {
            long requestId = requestCounter.incrementAndGet();

            getLogger().debug("Processing request #{} for pattern '{}': {} {}",
                    requestId, matchedPattern, request.getMethod(), request.getPath());

            // Create FlowFile from request
            FlowFile flowFile = createFlowFileFromRequest(session, request);

            // Transfer to success
            session.transfer(flowFile, REL_SUCCESS);
            session.commitAsync();

            getLogger().info("Successfully processed request #{} for pattern '{}': {} {}",
                    requestId, matchedPattern, request.getMethod(), request.getPath());

        } catch (Exception e) {
            getLogger().error("Failed to process request: {}", e.getMessage(), e);
            session.rollback();
        }
    }

    /**
     * Creates a FlowFile from a gateway request.
     * This should be called from a proper NiFi session context.
     */
    private FlowFile createFlowFileFromRequest(ProcessSession session, GatewayRequest request) {
        FlowFile flowFile = session.create();

        // Write request body as FlowFile content
        if (request.getBody() != null && request.getBody().length > 0) {
            flowFile = session.write(flowFile, out -> out.write(request.getBody()));
        }

        // Add HTTP attributes
        Map<String, String> attributes = new HashMap<>();
        attributes.put("http.method", request.getMethod());
        attributes.put("http.path", request.getPath());
        attributes.put("http.client.address", request.getClientAddress());
        attributes.put("http.timestamp", request.getTimestamp().toString());

        // Add query parameters
        for (Map.Entry<String, String> param : request.getQueryParameters().entrySet()) {
            attributes.put("http.query." + param.getKey(), param.getValue());
        }

        // Add path parameters
        for (Map.Entry<String, String> param : request.getPathParameters().entrySet()) {
            attributes.put("http.path." + param.getKey(), param.getValue());
        }

        // Add headers
        for (Map.Entry<String, String> header : request.getHeaders().entrySet()) {
            attributes.put("http.header." + header.getKey().toLowerCase(), header.getValue());
        }

        flowFile = session.putAllAttributes(flowFile, attributes);

        return flowFile;
    }
}